File: ruby/solver.rb
#!/usr/bin/env ruby
# Usage: ./solver.rb path/to/puzzle.txt

require 'set'

class Board
  attr_accessor :digits # a string holding the board state

  def initialize(digits)
    @digits = digits
  end

  def to_s
    digits.scan(/.{27}/).map do |x|
      x.scan(/.{9}/).map do |y|
        y.scan(/.{3}/).map do |z|
          z.split('').join(' ')
        end.join(' | ')
      end.join("\n")
    end.join("\n------+-------+------\n")
  end

  # returns set of numbers which can be written into position i of the sudoku board
  def options(i)
    Set.new('1'..'9') - peers(i)
  end

  def first_empty_position
    digits.index('.')
  end

  private

  def peers(i)
    row_peers(i) + column_peers(i) + box_peers(i)
  end

  # returns row peers of digit in the position i (i in 0..80)
  def row_peers(i)
    row_number = i / 9
    digits_on_positions(
      row_number * 9...(row_number + 1) * 9,
      i
    )
  end

  # returns column peers of element in the position i (i in 0..80)
  def column_peers(i)
    column_number = i % 9
    digits_on_positions(
      (0...9).map { |p| p * 9 + column_number },
      i
    )
  end

  # returns 3x3 box peers of element in the position i (i in 0..80)
  def box_peers(i)
    box_coordinates = ->(p) { [p / 27, (p % 9) / 3] } # returns a tuple [row 0..2, column 0..2], this is actually an array, no tuples in Ruby. Using a lambda as Ruby lacks nested methods too.
    digits_on_positions(
      (0..80).select { |p| box_coordinates.call(p) == box_coordinates.call(i) },
      i
    )
  end

  # returns a set containing digits on given board positions, without the digit on the current position
  def digits_on_positions(positions, current_position)
    positions
      .reject { |p| p == current_position }
      .map { |p| digits[p] }
      .reject { |digit| digit == '.' }
      .to_set
  end
end

# returns a Board instance containing a solved puzzle
def solve(board)
  position = board.first_empty_position
  return board unless position
  board.options(position).each do |option|
    solution = solve(Board.new(
      board.digits[0...position] + option + board.digits[position + 1..-1]
    ))
    return solution if solution
  end
  nil
end

# loads puzzle string (81 characters, 1..9 or dots for unknowns) from a file
def load_puzzle
  puzzle_filename = ARGV[0]
  raise 'No puzzle file given' unless puzzle_filename
  puzzle_str = File.read(puzzle_filename)
  puzzle_str.gsub(/[^0-9\.]/, '')
end

begin
  board = Board.new(load_puzzle)
  puts solve(board)
rescue RuntimeError => e
  puts e.message
rescue Errno::ENOENT
  puts 'Error loading the puzzle file'
end

File: python/solver.py
#!/usr/bin/env python3

"""Sudoku solver
Usage: ./solver.py path/to/puzzle
"""

import re
import sys

class Board():
    def __init__(self, digits):
        self.digits = digits

    def __str__(self):
        """__str__ works like to_s in Ruby or Stringer interface in Go"""
        def line_str(line):
            return (' | ').join(
                map(
                    lambda three_digit_str: (' ').join(list(three_digit_str)),
                    re.findall(r'.{3}' , line)
                )
            )

        def three_boxes_str(boxes):
            return ('\n').join(
                map(
                    line_str,
                    re.findall(r'.{9}' , boxes)
                )
            )

        return ('\n------+-------+------\n').join(
            map(
                three_boxes_str,
                re.findall(r'.{27}', self.digits)
            )
        )

    def options(self, i):
        """returns set of numbers which can be written into position i of the sudoku board"""
        return set([str(x) for x in range(1, 10)]) - self.peers(i)

    def first_empty_position(self):
        """returns the index of the first not-filled board element, None if not found"""
        position = self.digits.find('.')
        if position >= 0:
            return position

    def peers(self, i):
        return self.row_peers(i) | self.column_peers(i) | self.box_peers(i)

    def row_peers(self, i):
        """returns row peers of digit in the position i (i in 0..80)"""
        row_number = i // 9
        return self.digits_on_positions(
            range(row_number * 9, (row_number + 1) * 9),
            i
        )

    def column_peers(self, i):
        """returns column peers of element in the position i (i in 0..80)"""
        column_number = i % 9
        return self.digits_on_positions(
            [p * 9 + column_number for p in range(9)],
            i
        )

    def box_peers(self, i):
        """returns 3x3 box peers of element in the position i (i in 0..80)"""
        def box_coordinates(p):
            return (p // 27, (p % 9) // 3) # (row 0..2, column 0..2)
        return self.digits_on_positions(
            [p for p in range(81) if box_coordinates(p) == box_coordinates(i)],
            i
        )

    def digits_on_positions(self, positions, current_position):
        """returns a set containing digits on given board positions, without the digit on the current position"""
        return set(
            filter(
                lambda digit: digit != '.',
                [self.digits[p] for p in positions if p != current_position]
            )
        )

def solve(board):
    position = board.first_empty_position()
    if position is None:
        return board
    for option in board.options(position):
        solution = solve(Board(
            board.digits[:position] + str(option) + board.digits[position + 1:]
        ))
        if solution is not None:
            return solution
    return None

def load_puzzle():
    if len(sys.argv) < 2:
        sys.exit('No puzzle file given')
    puzzle_filename = sys.argv[1]
    with open(puzzle_filename) as file_object:
        puzzle_str = file_object.read()
        return re.sub(r'[^0-9\.]', '', puzzle_str)

def main():
    try:
        board = Board(load_puzzle())
        print(solve(board))
    except OSError as error:
        print('Error loading the puzzle file: {0}'.format(error.strerror))

main()
